hack/update-knative-components.go

```diff
@@ -0,0 +1,258 @@
+			evt = "v" + val
+		case knCtrPrefix:
+			ctr = "v" + val
+		}
+
+		// if all values are acquired, no need to continue
+		if srv != "" && evt != "" && ctr != "" {
+			break
+		}
+	}
+	return
+}
+
+// try updating the version of component named by "repo" via 'sed'
+func tryUpdateFile(prefix, newV, oldV string) (bool, error) {
+	fmt.Printf("> try updating %s(%s -> %s)...", prefix, oldV, newV)
+	quoteWrap := func(s string) string {
+		if !strings.HasPrefix(s, "\"") {
+			return "\"" + s + "\""
+		}
+		return s
+	}
+	if newV != oldV {
+		fmt.Println("updating")
+		cmd := exec.Command("sed", "-i", "-e", "s/"+prefix+quoteWrap(oldV)+"/"+prefix+quoteWrap(newV)+"/g", file)
+		err := cmd.Run()
+		if err != nil {
+			return false, fmt.Errorf("error while updating file: %s", err)
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// prepare branch for PR via git commands
+func prepareBranch(branchName string) error {
+	fmt.Println("> preparing branch")
+	cmd := exec.Command("bash", "-c", fmt.Sprintf(`
+		git config --local user.email "automation@knative.team" &&
+		git config --local user.name "Knative Automation" &&
+		git switch -c %s &&
+		git add %s &&
+		git commit -m "update components" &&
+		git push origin %s
+	`, branchName, file, branchName))
+	return cmd.Run()
+}
+
+// create a PR for the new updates
+func createPR(ctx context.Context, client *github.Client, title string, branchName string) error {
+	fmt.Print("> creating PR...")
+	body := fmt.Sprintf("%s\n/assign @gauron99", title)
+
+	newPR := github.NewPullRequest{
+		Title:               github.Ptr(title),
+		Base:                github.Ptr("main"),
+		Head:                github.Ptr(branchName),
+		Body:                github.Ptr(body),
+		MaintainerCanModify: github.Ptr(true),
+	}
+	pr, _, err := client.PullRequests.Create(ctx, "knative", "func", &newPR)
+	if err != nil {
+		fmt.Printf("PR looks like this:\n%#v\n", pr)
+		fmt.Printf("err: %s\n", err)
+		return err
+	}
+	fmt.Println("ready")
+	return nil
+}
+
+// returns true when PR with given title already exists in knative/func repo
+// otherwise false. Returns an error if occurred, otherwise nil.
+func prExists(ctx context.Context, c *github.Client, title string) (bool, error) {
+	opt := &github.PullRequestListOptions{State: "open"}
+	list, _, err := c.PullRequests.List(ctx, "knative", "func", opt)
+	if err != nil {
+		return false, fmt.Errorf("errror pulling PRs in knative/func: %s", err)
+	}
+	for _, pr := range list {
+		if pr.GetTitle() == title {
+			// gauron99 - currently cannot update already existing PR, shouldnt happen
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// ----------------------------------------------------------------------------
+// ----------------------------------- MAIN -----------------------------------
+// ----------------------------------------------------------------------------
+
+// entry function -- essentially "func main() for this file"
+func updateComponentVersions() error {
```

Comment:
Putting a "main" at the bottom of a file is a relic of a bygone era where symbols had to be defined before they were used, essentially forcing a file to be written upside-down.

In compiled languages, it is not necessary, allowing us to write code more like an essay, with the opening paragraph being the highest-level instruction set (or table of contents, if you like), which is then expanded upon in a depth-first-search way as the file continues, with ancillary/misc/util functions relegated to the bottom.

For readability, putting this function first in the file (just after  constants and global variables) is a great way to introduce a reader to the purpose of the code, slowly resolving into more detail as one reads on.